### PR TITLE
Added support for Commands in preparation for forcing Windows agent u…

### PR DESCRIPTION
**Pull request by jamesbrown7**
…pgrades

### CODE REVIEW

**Comment by jamesbrown7** [comment]
Closing while I work again on the unmarshalling stuff. Must have been good drugs or brain damage.

File: jcapi-commands.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+		resultsBlock, err2 := getJCCommandsFromResults(jcSysRec)
+		if err2 != nil {
+			err = fmt.Errorf("Could not get resultsBlock data, err='%s'", err2.Error())
+			return
+		}
+
+		for i, _ := range resultsBlock {
+			if resultsBlock[i].Id != "" {
+				commandList = append(commandList, resultsBlock[i])
+			}
```
Comment (ppg) [inline jcapi-commands.go:83]: Did you want to print a warning if the command has no Id?  Or is that an expected case?

File: jcapi-commands.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+//
+// Add or Update a command in place on JumpCloud
+//
+func (jc JCAPI) AddUpdateCommand(op JCOp, command JCCommand) (id string, err JCError) {
+	id, err = jc.HandleCommand(COMMAND_PATH, op, command)
+
+	return
+}
```
Comment (ppg) [inline jcapi-commands.go:117]: To keep the API consistent, I think I'd split this into `AddCommand` and `UpdateCommand` and have them pass the correct JCOp into `HandleCommand`. Otherwise I'd argue that you should check here that `op` is `Insert` or `Update`, and if not return an error.  But having two ways to call command crud in here doesn't seem worth while.

File: jcapi-commands.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+	commandResult := JCCommand{}
+
+	err = json.Unmarshal(result, &commandResult)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: Could not unmarshal result '%s', err='%s'", string(result), err.Error())
+	}
+
+	return commandResult.Id, nil
+}
```
Comment (ppg) [inline jcapi-commands.go:143]: This feels like a private worker function; do we want consumers to have access to this particular function, or just the more explicit crud actions?

File: jcapi-tags.go [hunk 4]
```diff
@@ -42,84 +51,59 @@ func GetTagNames(tags []JCTag) []string {
...
+		if err != nil {
+			err = fmt.Errorf("Could not unmarshal result '%s', err='%s'", string(result), err.Error())
+		}
 
-	builder = append(builder, buildJSONKeyValueBoolPair("externallyManaged", tag.ExternallyManaged))
-	builder = append(builder, buildJSONKeyValuePair("externalDN", tag.ExternalDN))
-	builder = append(builder, buildJSONKeyValuePair("externalSourceType", tag.ExternalSourceType))
+		tags = tagResults.Results
+	} else {
+		tags = append(tags, tag)
```
Comment (ppg) [inline jcapi-tags.go:72]: The append here, while correct, I think implies that `tags` might have other data, which it can't here.  I think I'd write this as `tags = []JCTag{tag}` so that its clear in one part of the if block you set it to `.Results` and in the other, a single item array.

File: jcapi-tags.go [hunk 5]
```diff
@@ -42,84 +51,59 @@ func GetTagNames(tags []JCTag) []string {
...
+	tagList, err = getJCTagsFromResults(result)
+	if err != nil {
+		return nil, fmt.Errorf("ERROR: Could not get tags from results, err='%s'", err.Error())
 	}
 
-	if tagData["externalSourceType"] != nil {
-		tag.ExternalSourceType = getStringOrNil(tagData["externalSourceType"].(interface{}))
-	}
+	return
 }
```
Comment (ppg) [inline jcapi-tags.go:91]: This one seems like potentially a private worker function too?

File: jcapi-tags.go [hunk 6]
```diff
@@ -127,15 +111,16 @@ func (jc JCAPI) GetAllTags() (tagList []JCTag, err JCError) {
...
-		err = fmt.Errorf("ERROR: Could not get tag by name/id '%s', err='%s'", tagName, err)
+		err = fmt.Errorf("ERROR: Could not get tags by name for '%s', url='%s', err='%s'", tagName, url, err.Error())
+		return
 	}
 
-	if retVal != nil {
-		jc.getTagFieldsFromInterface(retVal.(map[string]interface{}), &tag)
+	if len(tags) > 0 {
+		tag = tags[0]
 	}
```
Comment (ppg) [inline jcapi-tags.go:124]: Might want to at least log a warning if `len(tags) > 1` to help people see if/when that issue occurs.

File: jcapi-tags.go [hunk 7]
```diff
@@ -127,15 +111,16 @@ func (jc JCAPI) GetAllTags() (tagList []JCTag, err JCError) {
...
 
 	return
```
Comment (ppg) [inline jcapi-tags.go:126]: What is the desired behavior if one calls `GetTagByName` with a name that doesn't exist? Right now I think it returns an error that has some context around '404', but if `GetTagsByUrl` could return no err but an empty tags, you could exit from this function with `nil, nil` as well; if we want 404 for missing tag, we should probably instead check `len(tags) == 0` and return an error, otherwise return `tags[0]`.

File: jcapi-tags.go [hunk 8]
```diff
@@ -145,23 +130,33 @@ func (jc JCAPI) GetTagByName(tagName string) (tag JCTag, err JCError) {
 // Add or Update a tag in place on JumpCloud
 //
 func (jc JCAPI) AddUpdateTag(op JCOp, tag JCTag) (tagId string, err JCError) {
-	data, err := tag.MarshalJSON()
+	data, err := json.Marshal(tag)
```
Comment (ppg) [inline jcapi-tags.go:133]: Just a note so its captured for others.  If your struct implements `MarshalJSON`, `json.Marshal` should actually call that itself, no need to ever directly call it.

File: jcapi-tags.go [hunk 9]
```diff
@@ -173,7 +168,7 @@ func (jc JCAPI) AddUpdateTag(op JCOp, tag JCTag) (tagId string, err JCError) {
 }
```
Comment (ppg) [inline jcapi-tags.go:168]: Ditto, semantically can we split into `AddTag` and `UpdateTag`?

File: jcapi.go [hunk 10]
```diff
@@ -186,6 +187,38 @@ func (jc JCAPI) Do(op, url string, data []byte) (interface{}, JCError) {
...
+	}
+
+	buffer, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("ERROR: Could not read the response body, err='%s'", err)
+	}
+
+	return buffer, err
+}
+
```
Comment (ppg) [inline jcapi.go:221]: Not sure about feelings on TODO comments, but it'd be nice to note somewhere that `Do` should be avoided in favor of `DoBytes` for any new code until a point in time at which they are collapsed into one function.

File: jcapi.go [hunk 11]
```diff
@@ -240,26 +273,31 @@ func (jc JCAPI) extractStringArray(input []interface{}) []string {
 	return returnVal
 }
```
Comment (ppg) [inline jcapi.go:274]: If you're touching this code anyway, like `getStringOrNil` below, you can declare `returnVal` in the method signatures return arguments, and avoid the var line here.

Comment (ppg) [inline jcapi.go:274]: Although, after having said that, you might also want to pre-allocate the return string array to the capacity of `len(input)`.  But I could see the argument to leave those to a different PR/refactor.

File: jcapi.go [hunk 12]
```diff
@@ -315,3 +353,61 @@ func FindObject(sourceArray []interface{}, fieldName string, compareData interfa
...
+			return
+		}
+
+		if r.Match([]byte(s.String())) {
+			index = fieldIndex
+			return
+		}
+	}
+
+	index = OBJECT_NOT_FOUND
```
Comment (ppg) [inline jcapi.go:411]: I thought you had this at the top before?  I like that better, as there are numerous exit paths where err is set, but index will be 0; I think it'd be more robust if index was always == OBJECT_NOT_FOUND when err is not nil.

File: jcapi_test.go [hunk 13]
```diff
@@ -498,12 +498,17 @@ func TestRadiusServer(t *testing.T) {
...
+		switch radServer.Id {
+		case rs1.Id:
+			if radServer.ToString() != rs1.ToString() {
+				t.Fatalf("radServer='%s' - rs1='%s' - string compare failed", radServer.ToString(), rs1.ToString())
+			}
+		case rs3.Id:
+			if radServer.ToString() != rs3.ToString() {
+				t.Fatalf("radServer='%s' - rs3='%s' - string compare failed", radServer.ToString(), rs3.ToString())
+			}
+		}
```
Comment (ppg) [inline jcapi_test.go:511]: This might be cleaned up by first generating actual and expected string arrays of rad servers, then deep reflecting they're equal. Alternatively, even if one doesn't want to use goconvey's (our current TDD/BDD framework of choice) full syntax, they can still pull in the underlying assertion helpers they use and use those: https://github.com/smartystreets/assertions/blob/master/equality.go#L139; then you could write something like:

```
message := assetions.ShouldResemble(radserves, []RadServer{r1, r3})
if message != "" {
  t.Fatal(message)
}
```
